docs/api/system_and_configuration/configuration_info.md
## Configuration Info [/config_info]
## Configuration Info [GET]
Configuration information gives you detailed information about configuration of Judge0.
This configuration can be changed through [judge0.conf](https://github.com/judge0/judge0/blob/master/judge0.conf)
file by admin who hosts Judge0 instance.

This configuration gives every admin a flexibility to configure Judge0 according to server abilities and needs. It also gives users
insight on some *default configuration values* which are used when their programs are run.

Each of these *configuration variables* have *default values* which we consider as recommended in case you are not sure should you change them.

We will refer to *default values* as values which Judge0 automatically assigns to each of these *configuration variables*,
if admin didn't set them. For example, default value of *configuration variable* `cpu_time_limit` is `2`.

|#    |Name|Type |Unit |Description|Default Value|
|:---:|:---|:---:|:---:|:----------|:------------|
|1|`enable_wait_result`|boolean||If enabled user can request to synchronously wait for submission result on submission create.|true|
|2|`enable_compiler_options`|boolean||If enabled user can set `compiler_options`.|true|
|3|`allowed_languages_for_compile_options`|list of strings||Languages for which setting compiler options is allowed.|Empty, i.e. for all languages it is allowed to set compiler options.|
|4|`enable_command_line_arguments`|boolean||If enabled user can set `command_line_arguments`.|true|
|5|`enable_submission_delete`|boolean||If enabled authorized user can [delete a submission](#submissions-submission-delete).|false|
|6|`max_queue_size`|integer||Maximum number of submissions that can wait in queue.|100|
|7|`cpu_time_limit`|float|second|Default runtime limit for every program (in seconds). Decimal numbers are allowed. Time in which the OS assigns the processor to different tasks is not counted.|2|
|8|`cpu_extra_time`|float|second|When a time limit is exceeded, wait for extra time, before killing the program. This has the advantage that the real execution time is reported, even though it slightly exceeds the limit.|0.5|
|9|`wall_time_limit`|float|second|Limit wall-clock time in seconds. Decimal numbers are allowed. This clock measures the time from the start of the program to its exit, for an external event. We recommend to use `cpu_time_limit` as the main limit, but set `wall_time_limit` to a much higher value as a precaution against sleeping programs.|5|
|10|`memory_limit`|integer|kilobyte|Limit address space of the program in kilobytes.|128000|
|11|`stack_limit`|integer|kilobyte|Limit process stack in kilobytes.|64000|
|12|`max_processes_and_or_threads`|integer||Maximum number of processes and/or threads program can create.|60|
|13|`enable_per_process_and_thread_time_limit`|boolean||If `true` then `cpu_time_limit` will be used as per process and thread.|false|
|14|`enable_per_process_and_thread_memory_limit`|boolean||If `true` then `memory_limit` will be used as per process and thread.|true|
|15|`max_file_size`|integer|kilobyte|Limit size of files created (or modified) by the program.|1024|
|16|`allow_enable_network`|boolean||If enabled user can set `enable_network`.|true|
|17|`enable_network`|boolean||If enabled program will have network access.|true|
|18|`number_of_runs`|integer||Run each program this many times and take average of time and memory.|1|

*Default configuration value* for each variable is given to you as response of this API call. For example, *default configuration value*
for variable `cpu_extra_time` might be `2`, and if admin didn't set this, then it is `0.5` (*default value*).
This means that admin set `cpu_extra_time` *configuration variable* to value `2` and we say it is now *default configuration value* for this
variable `cpu_extra_time`.

Every [submission](#submissions-submission) can change each of the configuration variables according to its needs. For example,
user might create submission which has `cpu_time_limit` of `5` seconds. For security reasons we need to limit values of each of these
configuration variables. For example, we don't want user to create a submission which has `cpu_time_limit` of `100000` seconds.

For this security reason we are introducing *limit configuration variables* for each *configuration variable*.

|#    |Name|Type |Unit |Description|Default Value|
|:---:|:---|:---:|:---:|:----------|:------------|
|1|`max_cpu_time_limit`|float|second|Maximum custom `cpu_time_limit`|15|
|2|`max_cpu_extra_time`|float|second|Maximum custom `cpu_extra_time`|2|
|3|`max_wall_time_limit`|float|second|Maximum custom `wall_time_limit`|20|
|4|`max_memory_limit`|integer|kilobyte|Maximum custom `memory_limit`|256000|
|5|`max_stack_limit`|integer|kilobyte|Maximum custom `stack_limit`|128000|
|6|`max_max_processes_and_or_threads`|integer||Maximum custom `max_processes_and_or_threads`|120|
|7|`allow_enable_per_process_and_thread_time_limit`|boolean||If `false` user won't be able to set `enable_per_process_and_thread_time_limit` to `true`|true|
|8|`allow_enable_per_process_and_thread_memory_limit`|boolean||If `false` user won't be able to set `enable_per_process_and_thread_memory_limit` to `true`|true|
|9|`max_max_file_size`|integer|kilobyte|Maximux custom `max_file_size`|4096|
|10|`max_number_of_runs`|integer||Maximum custom `number_of_runs`|20|

For example, `max_cpu_time_limit` with value `20` means that user cannot create new submission which has `cpu_time_limit` greater than `20`.

+ Response 200 (application/json)
    {
        "enable_wait_result": true,
        "enable_compiler_options": true,
        "allowed_languages_for_compile_options": [],
        "enable_command_line_arguments": true,
        "enable_submission_delete": false,
        "max_queue_size": 100,
        "cpu_time_limit": 2,
        "max_cpu_time_limit": 15,
        "cpu_extra_time": 0.5,
        "max_cpu_extra_time": 2,
        "wall_time_limit": 5,
        "max_wall_time_limit": 20,
        "memory_limit": 128000,
        "max_memory_limit": 256000,
        "stack_limit": 64000,
        "max_stack_limit": 128000,
        "max_processes_and_or_threads": 60,
        "max_max_processes_and_or_threads": 120,
        "enable_per_process_and_thread_time_limit": false,
        "allow_enable_per_process_and_thread_time_limit": true,
        "enable_per_process_and_thread_memory_limit": true,
        "allow_enable_per_process_and_thread_memory_limit": true,
        "max_file_size": 1024,
        "max_max_file_size": 4096,
        "number_of_runs": 1,
        "max_number_of_runs": 20
    }

docs/api/system_and_configuration/system_and_configuration.md
# Group System and Configuration
<!-- include(system_info.md) -->
<!-- include(configuration_info.md) -->

CHANGELOG.md
This is the Changelog for Judge0 CE. The Changelog for Judge0 Extra CE can be found [here](https://github.com/judge0/judge0/blob/extra/CHANGELOG.md). Learn more about their difference [here](https://github.com/judge0/judge0#flavors).

# v1.13.0 (2021-03-10)
Huge thanks to [**Filtered**](https://www.filtered.ai) for sponsoring this release.

## New Features
- Added support for `enable_network` configuration flag. With the new `ALLOW_ENABLE_NETWORK` configuration variable, usage of this flag can be permitted or denied. Furthermore, with the new `ENABLE_NETWORK` configuration variable the default value of this flag can be set for every submission.
    - Commits: [@62a00520](https://github.com/judge0/judge0/commit/62a00520692c2b7a8bf0f54528aee8fbbf864f36)
- Added `USE_DOCS_AS_HOMEPAGE` configuration variable, which allows you to show or hide Judge0 homepage. **Now, by default, Judge0 homepage is empty** and does not show the API documentation. However, you can still access the API documentation via `/docs`.
    - Issues: [#257](https://github.com/judge0/judge0/issues/257)

## Bug Fixes
- Fixed the bug where the wrong number of workers would be reported via `/workers`.
	- Issues: [#256](https://github.com/judge0/judge0/issues/256)

## Security Fixes
- **HIGH** Fixed a security bug where certain submission configuration settings would allow the user to run a program that would run infinetly long. With this fix `wall_time_limit` **must** be at least 1 second.
	- Commits: [@fce8d97a](https://github.com/judge0/judge0/commit/fce8d97ae765e87c4584115fcb73bd288d64d815)

## Other Changes
- Allow setting the `max_file_size` and `stack_limit` to 0 kB.
	- Issues: [#242](https://github.com/judge0/judge0/issues/242)
	- Commits: [@bf7c1284](https://github.com/judge0/judge0/commit/bf7c1284fe756d348154546ad8a6280ce3af5eb2) [@5a4b79ab](https://github.com/judge0/judge0/commit/5a4b79abd19c20b7b26ef70cf862da25f5fd8008)
- Allow setting the `cpu_time_limit` and `cpu_extra_time` to 0 seconds.
	- Commits: [@fce8d97a](https://github.com/judge0/judge0/commit/fce8d97ae765e87c4584115fcb73bd288d64d815)
- Updated to a port 2358 as a new **default port for Judge0** as a online code execution service.
	- Issues: [#205](https://github.com/judge0/judge0/issues/205)
	- Commits: [@ea9c7c97](https://github.com/judge0/judge0/commit/ea9c7c975e0697d66ada910ec7806d3c28ad4a03)
- Updated the default number of Judge0 Workers, those that acutally run the user's code, to the 2*[`nproc`](https://linux.die.net/man/1/nproc). This has been shown as a good choice for general purpose use-case.
	- Commits: [@113d9c74](https://github.com/judge0/judge0/commit/113d9c74b4659cf6707e748eb6e04701ce67d3f0)
- Updated the default number of Rails threads to `nproc` and Rails processes to 2. This has been shown as a good choice for general purpose use-case.
	- Commits: [@113d9c74](https://github.com/judge0/judge0/commit/113d9c74b4659cf6707e748eb6e04701ce67d3f0)
- Updated Let's Encrypt Docker image for deployment with HTTPS.
	- Commits: [@86b7f8e8](https://github.com/judge0/judge0/commit/86b7f8e8d0154ab5044db91c537872ea8deb7343)
- Updated Nginx proxy Docker image that is used in development setup and deployment with HTTPS.
	- Commits: [@83f5b175](https://github.com/judge0/judge0/commit/83f5b175d0925fe21374e9c70d7fbb4e6941fa9c)
- Fixed documentation typos. Thank you @balababa.
	- Pull Requests: [#245](https://github.com/judge0/judge0/pull/245)
- Updated base image to `judge0/compilers:1.4.0` which uses updated Isolate to [@ad39cc4d](https://github.com/judge0/isolate/commit/ad39cc4d0fbb577fb545910095c9da5ef8fc9a1a).

## Deployment Procedure
Judge0 is collecting telemetry data to help understand how to improve the product and to better understand how Judge0 is used in various production environments. Read more about telemetry [here](https://github.com/judge0/judge0/blob/v1.13.0/TELEMETRY.md).

Please note that Judge0 has only been tested on **Linux** and **macOS**, and might not work on Windows, thus we do not provide support for it.

### With HTTP
1. Install [Docker](https://docs.docker.com) and [Docker Compose](https://docs.docker.com/compose).
2. Download and extract the release archive:
```
wget https://github.com/judge0/judge0/releases/download/v1.13.0/judge0-v1.13.0.zip
unzip judge0-v1.13.0.zip
```

3. Run all services and wait a few seconds until everything is initialized:
```
cd judge0-v1.13.0
docker-compose up -d db redis
sleep 10s
docker-compose up -d
sleep 5s
```

4. Your instance of Judge0 CE v1.13.0 is now available at `http://<IP ADDRESS OF YOUR SERVER>:2358`.

### With HTTPS (SSL/TLS)
1. Install [Docker](https://docs.docker.com) and [Docker Compose](https://docs.docker.com/compose).
2. Download and extract the release archive:
```
wget https://github.com/judge0/judge0/releases/download/v1.13.0/judge0-v1.13.0-https.zip
unzip judge0-v1.13.0-https.zip
```

3. Change directory to `judge0-v1.13.0-https`:
```
cd judge0-v1.13.0-https
```
4. Edit `docker-compose.yml` and change variables `VIRTUAL_HOST`, `LETSENCRYPT_HOST` and `LETSENCRYPT_EMAIL`.
5. Run all services and wait a few seconds until everything is initialized:
```
docker-compose up -d db redis
sleep 10s
docker-compose up -d
sleep 5s
```

6. Your instance of Judge0 CE v1.13.0 is now available at `https://<YOUR DOMAIN>`.


# v1.12.0 (2020-10-18)
## New Features
- Added support for invalidating statistics cache using the `invalidate_cache` query parameter.
    - Commits: [@561c2a30](https://github.com/judge0/judge0/commit/561c2a309d98680801628fc2d79ea17e885e1d66)
- Added rake task `judge0:run_in_queue` to manually run submissions that might have stuck in the `In Queue` state.
    - Commits: [@c3b87d23](https://github.com/judge0/judge0/commit/c3b87d232a8e922b5ae04f5afdb12fc49b2ec003)
- Added support for specifying `CALLBACKS_MAX_TRIES` and `CALLBACKS_TIMEOUT`. This is a global configuration that applies to all submissions that have defined webhooks (HTTP callbacks).

app/helpers/config.rb
module Config
  # For more info read:
  # https://github.com/judge0/judge0/blob/master/judge0.conf

  MAINTENANCE_MODE = ENV["MAINTENANCE_MODE"] == "true"
  ENABLE_WAIT_RESULT = ENV["ENABLE_WAIT_RESULT"] != "false"
  ENABLE_COMPILER_OPTIONS = ENV["ENABLE_COMPILER_OPTIONS"] != "false"
  ALLOWED_LANGUAGES_FOR_COMPILER_OPTIONS = ENV["ALLOWED_LANGUAGES_FOR_COMPILER_OPTIONS"].to_s.strip.split
  ENABLE_COMMAND_LINE_ARGUMENTS = ENV["ENABLE_COMMAND_LINE_ARGUMENTS"] != "false"
  ENABLE_SUBMISSION_DELETE = ENV["ENABLE_SUBMISSION_DELETE"] == "true"
  ENABLE_CALLBACKS = ENV["ENABLE_CALLBACKS"] != "false"
  CALLBACKS_MAX_TRIES = (ENV["CALLBACKS_MAX_TRIES"].presence || 3).to_i
  CALLBACKS_TIMEOUT = (ENV["CALLBACKS_TIMEOUT"].presence || 5).to_f
  ENABLE_ADDITIONAL_FILES = ENV["ENABLE_ADDITIONAL_FILES"] != "false"
  MAX_QUEUE_SIZE = (ENV["MAX_QUEUE_SIZE"].presence || 100).to_i
  CPU_TIME_LIMIT = (ENV["CPU_TIME_LIMIT"].presence || 15).to_f
  MAX_CPU_TIME_LIMIT = (ENV["MAX_CPU_TIME_LIMIT"].presence || 20).to_f
  CPU_EXTRA_TIME = (ENV["CPU_EXTRA_TIME"].presence || 1).to_f
  MAX_CPU_EXTRA_TIME = (ENV["MAX_CPU_EXTRA_TIME"].presence || 5).to_f
  WALL_TIME_LIMIT = (ENV["WALL_TIME_LIMIT"].presence || 20).to_f
  MAX_WALL_TIME_LIMIT = (ENV["MAX_WALL_TIME_LIMIT"].presence || 25).to_f
  MEMORY_LIMIT = (ENV["MEMORY_LIMIT"].presence || 128000).to_i # in KB
  MAX_MEMORY_LIMIT = (ENV["MAX_MEMORY_LIMIT"].presence || 512000).to_i
  STACK_LIMIT = (ENV["STACK_LIMIT"].presence || 64000).to_i # in KB
  MAX_STACK_LIMIT = (ENV["MAX_STACK_LIMIT"].presence || 128000).to_i
  MAX_PROCESSES_AND_OR_THREADS = (ENV["MAX_PROCESSES_AND_OR_THREADS"].presence || 100).to_i
  MAX_MAX_PROCESSES_AND_OR_THREADS = (ENV["MAX_MAX_PROCESSES_AND_OR_THREADS"].presence || 120).to_i
  ENABLE_PER_PROCESS_AND_THREAD_TIME_LIMIT = ENV["ENABLE_PER_PROCESS_AND_THREAD_TIME_LIMIT"] == "true"
  ALLOW_ENABLE_PER_PROCESS_AND_THREAD_TIME_LIMIT = ENV["ALLOW_ENABLE_PER_PROCESS_AND_THREAD_TIME_LIMIT"] != "false"
  ENABLE_PER_PROCESS_AND_THREAD_MEMORY_LIMIT = ENV["ENABLE_PER_PROCESS_AND_THREAD_MEMORY_LIMIT"] == "true"
  ALLOW_ENABLE_PER_PROCESS_AND_THREAD_MEMORY_LIMIT = ENV["ALLOW_ENABLE_PER_PROCESS_AND_THREAD_MEMORY_LIMIT"] != "false"
  MAX_FILE_SIZE = (ENV["MAX_FILE_SIZE"].presence || 5120).to_i
  MAX_MAX_FILE_SIZE = (ENV["MAX_MAX_FILE_SIZE"].presence || 20480).to_i
  NUMBER_OF_RUNS = (ENV["NUMBER_OF_RUNS"].presence || 1).to_i
  MAX_NUMBER_OF_RUNS = (ENV["MAX_NUMBER_OF_RUNS"].presence || 20).to_i
  REDIRECT_STDERR_TO_STDOUT = ENV["REDIRECT_STDERR_TO_STDOUT"] == "true"
  MAX_EXTRACT_SIZE = (ENV["MAX_EXTRACT_SIZE"].presence || 10240).to_i
  ENABLE_BATCHED_SUBMISSIONS = ENV["ENABLE_BATCHED_SUBMISSIONS"] != "false"
  MAX_SUBMISSION_BATCH_SIZE = (ENV["MAX_SUBMISSION_BATCH_SIZE"].presence || 20).to_i
  SUBMISSION_CACHE_DURATION = (ENV["SUBMISSION_CACHE_DURATION"].presence || 1).to_f
  USE_DOCS_AS_HOMEPAGE = ENV["USE_DOCS_AS_HOMEPAGE"] == "true"
  ALLOW_ENABLE_NETWORK = ENV["ALLOW_ENABLE_NETWORK"] != "false"
  ENABLE_NETWORK = ENV["ENABLE_NETWORK"] != "false" # Enabled by default for Judge0 Extra CE.

  def self.config_info
    @@default_confg ||= {
      "maintenance_mode": MAINTENANCE_MODE,
      "enable_wait_result": ENABLE_WAIT_RESULT,
      "enable_compiler_options": ENABLE_COMPILER_OPTIONS,
      "allowed_languages_for_compile_options": ALLOWED_LANGUAGES_FOR_COMPILER_OPTIONS,
      "enable_command_line_arguments": ENABLE_COMMAND_LINE_ARGUMENTS,
      "enable_submission_delete": ENABLE_SUBMISSION_DELETE,
      "enable_callbacks": ENABLE_CALLBACKS,
      "callbacks_max_tries": CALLBACKS_MAX_TRIES,
      "callbacks_timeout": CALLBACKS_TIMEOUT,
      "enable_additional_files": ENABLE_ADDITIONAL_FILES,
      "max_queue_size": MAX_QUEUE_SIZE,
      "cpu_time_limit": CPU_TIME_LIMIT,
      "max_cpu_time_limit": MAX_CPU_TIME_LIMIT,
      "cpu_extra_time": CPU_EXTRA_TIME,
      "max_cpu_extra_time": MAX_CPU_EXTRA_TIME,
      "wall_time_limit": WALL_TIME_LIMIT,
      "max_wall_time_limit": MAX_WALL_TIME_LIMIT,
      "memory_limit": MEMORY_LIMIT,
      "max_memory_limit": MAX_MEMORY_LIMIT,
      "stack_limit": STACK_LIMIT,
      "max_stack_limit": MAX_STACK_LIMIT,
      "max_processes_and_or_threads": MAX_PROCESSES_AND_OR_THREADS,
      "max_max_processes_and_or_threads": MAX_MAX_PROCESSES_AND_OR_THREADS,
      "enable_per_process_and_thread_time_limit": ENABLE_PER_PROCESS_AND_THREAD_TIME_LIMIT,
      "allow_enable_per_process_and_thread_time_limit": ALLOW_ENABLE_PER_PROCESS_AND_THREAD_TIME_LIMIT,
      "enable_per_process_and_thread_memory_limit": ENABLE_PER_PROCESS_AND_THREAD_MEMORY_LIMIT,
      "allow_enable_per_process_and_thread_memory_limit": ALLOW_ENABLE_PER_PROCESS_AND_THREAD_MEMORY_LIMIT,
      "max_file_size": MAX_FILE_SIZE,
      "max_max_file_size": MAX_MAX_FILE_SIZE,
      "number_of_runs": NUMBER_OF_RUNS,
      "max_number_of_runs": MAX_NUMBER_OF_RUNS,
      "redirect_stderr_to_stdout": REDIRECT_STDERR_TO_STDOUT,
      "max_extract_size": MAX_EXTRACT_SIZE,
      "enable_batched_submissions": ENABLE_BATCHED_SUBMISSIONS,
      "max_submission_batch_size": MAX_SUBMISSION_BATCH_SIZE,
      "submission_cache_duration": SUBMISSION_CACHE_DURATION,
      "use_docs_as_homepage": USE_DOCS_AS_HOMEPAGE,
      "allow_enable_network": ALLOW_ENABLE_NETWORK,
      "enable_network": ENABLE_NETWORK
    }
  end
end
